### PR TITLE
Fixed KT-39868: Allow @JvmStatic on protected properties in companion objects.

### PR DESCRIPTION
**Pull request by BomBardyGamer**
Specifically, I have changed both the FIR and the old frontend checkers to verify that the declaration visibility is not protected. In the FIR version, we fast exit if the declaration is any member declaration with a visibility of protected. In the non-FIR version, we incorporate the visibility check as part of the existing statements, as unlike in the FIR version, the declaration is required to be a property declaration.

I am not sure if this is the way that this should be fixed. It would probably be nicer if we could have protected constants work without JvmStatic, but that's for another time.

Contribution checklist:
* Link to issue: [KT-39868](https://youtrack.jetbrains.com/issue/KT-39868/False-positive-JvmStatic-annotation-is-useless-for-const-of-JvmField-properties-for-const-inside-a-companion-object)
* Changes made are the exact fix that is required.
* Changes have been explained at the beginning of the PR.
* I cannot build the entire project as my machine is not powerful enough (I have 32 GB RAM and a Ryzen 7 3700X, but the compiler is a very large project), but I am certain that my changes would pass a build if it was to be ran on, for example, TeamCity.
* The four tests that were added to (can be found by looking at changed files) all successfully pass on my local machine. Note that special tests that require JDK 17 or 11 to be set did not pass, however the behaviour of these is not affected.
* There are no merge commits in the PR.

### CODE REVIEW

**Review by mglukhikh — CHANGES_REQUESTED** [review]
Apologies but it's not enough to add only a diagnostic test. To ensure that the code works in runtime, you have to add also a black box test `compiler/testData/codegen/box`. You should add a comment `// TARGET_BACKEND: JVM` at the beginning, because `@JvmStatic` is the JVM-only thing. Any black box test should contain `fun box(): String` function returning "OK" in case of success. Of course the related feature (protected member of a companion object) should be used in runtime. 

To be honest, I don't remember details of this situation right now, but definitely such protected members were forbidden due to some reason, not just to forbid something, and I suspect that your black box test will fail. But anyway you should check it first.

**Comment by BomBardyGamer** [comment]
Hey, I've added the black box test as you asked, and I've ran all four generated tests that changed, and all four of them pass successfully. I'm not sure my test is the best, but it's really all I can think of without violating the access restrictions of `protected` members.

**Review by mglukhikh — APPROVED** [review]

File: compiler/testData/codegen/box/jvmStatic/kt39868.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+open class A {
+
+    companion object {
+
+        @JvmStatic
+        protected const val x = 1
+
+        @JvmStatic
+        @JvmField
+        protected val z = 1
```
Comment (udalov) [inline compiler/testData/codegen/box/jvmStatic/kt39868.kt:13]: Please make this a `var` and check that changing its value works

**Review by udalov — CHANGES_REQUESTED** [review]

**Comment by udalov** [comment]
Since this makes red code green, we need to allow this only if a language feature is enabled. You can add an entry to [`LanguageFeature`](https://github.com/JetBrains/kotlin/blob/396ac6f5e3f467a4c9e5307647eb8db3448e76e9/compiler/util/src/org/jetbrains/kotlin/config/LanguageVersionSettings.kt#L13) and use it in the checkers. It's important because a feature can only be enabled in a major version, e.g. 1.9, and `-language-version 1.8` should restore the old compiler behavior, exactly as it was in Kotlin compiler of version 1.8. Please also make two diagnostic tests: one where the feature is disabled and thus the error is still reported, and the other where it's enabled and there's no error. We use the `// !LANGUAGE` directive for that, see e.g. [before](https://github.com/JetBrains/kotlin/blob/59663f47520985d2324cb0a93879198e81fa4053/compiler/testData/diagnostics/tests/BreakContinueInWhen_before.kt) & [after](https://github.com/JetBrains/kotlin/blob/59663f47520985d2324cb0a93879198e81fa4053/compiler/testData/diagnostics/tests/BreakContinueInWhen_after.kt).

**Comment by BomBardyGamer** [comment]
Thanks for your comments. I've attempted to adjust my test to try reassigning the value, however it seems that the `@JvmStatic` annotation doesn't get applied to the synthetic setter generated (the one with the `<set-{name}>` name), but I'm not entirely sure what I should do.

If I add an explicit `@set:JvmStatic` to the variable, the test passes, as `@JvmStatic` is applied explicitly to the setter, but I don't think this is the sort of behaviour we should require. Though the alternatives are to either detect the synthetic setter in the call checker for `SUBCLASS_CANT_CALL_COMPANION_PROTECTED_NON_STATIC` (which is what's causing the issue), applying `@JvmStatic` to generated setters of properties with `@JvmStatic` anyway (which is what I thought it would've done already, I'm not sure if this is a bug or not), or I could just leave the behaviour as-is and require `@set:JvmStatic` to be applied to any property, but as I said, I don't think this is a good idea, as it's unnecessary boilerplate, and also introduces another implementation detail in to the language (or so it seems).

What do you think? I will push the revised (broken) test anyway, so you can explore it for yourself if you wish.

**Comment by udalov** [comment]
I'll close this PR because this needs to be resolved differently (see my comment above). I'll investigate it myself.

**Pull request closed without merging**